### PR TITLE
Shorten share URLs

### PR DESCRIPTION
**Pull request by gregdingle**
# Description

This shortens the long URLs on the visualization pages. 

image

Most of the work is handled by the `shortener` gem. 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix a feature that knowingly causes existing functionality to not work as expected)

# Test and Reproduce
Outline the steps to test or reproduce the PR here.

1. open a heatmap / report page
2. customize
3. hit share
5. paste the short url into a new tab
6. see the same vis state

## Impacted Areas in Application
List general components of the application that this PR will affect:

Which pages?

- [ ] Login Screen
- [ ] Single Upload Page
- [ ] Batch Upload Page
- [ ] All Projects Page
- [ ] Sample Details Page
- [X] Report Page

### CODE REVIEW

File: db/migrate/20190220191436_create_shortened_urls_table.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      t.string :unique_key, limit: 10, null: false
+
+      # a category to help categorize shortened urls
+      t.string :category
+
+      # how many times the link has been clicked
+      t.integer :use_count, null: false, default: 0
+
+      # valid until date for expirable urls
+      t.datetime :expires_at
```
Comment (tfrcarvalho) [inline db/migrate/20190220191436_create_shortened_urls_table.rb]: Cool! How are expired urls cleaned? Is there a rake task or something for that?

File: app/assets/src/helpers/url.js [hunk 2]
```diff
@@ -27,3 +29,9 @@ export const getURLParamString = params => {
     .map(pair => pair.join("="))
     .join("&");
 };
+
+export const copyShortUrlToClipboard = async url => {
+  url = url || window.location.href;
+  const shortUrl = await shortenUrl(url);
+  copy(window.location.origin + "/" + shortUrl.unique_key);
```
Comment (tfrcarvalho) [inline app/assets/src/helpers/url.js]: Form the snapshot noticed this unique is a little short (4 chars). Can we increase the size a bit?

File: db/migrate/20190220191436_create_shortened_urls_table.rb (hunk 1 above)
Comment (gregdingle) [reply db/migrate/20190220191436_create_shortened_urls_table.rb]: Expiry is optional on creation of a short url. Do you think we should add it? I didn't see any reason why.

File: app/assets/src/helpers/url.js (hunk 2 above)
Comment (gregdingle) [reply app/assets/src/helpers/url.js]: The default is 5 chars, which gives us 36^5 = 60M keys. I can configure it now, or we can wait until it is a problem and add more chars. What do you think?

Comment (tfrcarvalho) [reply app/assets/src/helpers/url.js]: If the change is backwards-compatible, its fine...
But if we are not expiring I would add a 2/3 more characters.

File: db/migrate/20190220191436_create_shortened_urls_table.rb (hunk 1 above)
Comment (tfrcarvalho) [reply db/migrate/20190220191436_create_shortened_urls_table.rb]: Just clean-up... I guess we can do it, if it becomes very popular.

**Pull request merged**